### PR TITLE
Update OncoTree-Mapping-Tool.md

### PR DESCRIPTION
**Pull request by ritikakundra**
- Updating with Ontology mapping section
- Added the ontology_to_ontology_mapping_tool

### CODE REVIEW

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:3]: I suggest making the name a little more specific. To be parallel with the other tool we have, perhaps this could be called the "Ontology to Ontology Mapping tool"? Other options: "External Ontology Mapping Tool", "Cross Ontology Mapping Tool", "Inter-Ontology Mapping Tool"

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:194]: This sounds good .. but this PR should also contain the script as well (ontology_mapping_tool.py) (or .. ontology_to_ontology_mapping_tool.py) before it gets merged.

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:200]: script rename suggested above

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:205]: What happens if the source file contains 2 or 3 of these fields? We should know what the script behavior should be and let users know here. Is there a preset order of preference? Or maybe we need a fourth command line argument "-s | --source-code" ?

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:207]: here you say "codes" .. but below you say that only one code can be accepted. Maybe this should be "code"?

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:212]: I'm not sure what "one or more columns" means ... don't all non-empty source files have one or more columns? Maybe this can be dropped.  Also "atleast" --> "at least"

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:213]: I would be a little less definite about extending the tool to support multiple mappings (unless we have a delivery date in mind). So I would suggest something like "In the future, we hope to extend the tool to support mapping to multiple ontology systems", or "We plan to extend the tool to support mapping to multiple ontology systems."
Also ... wouldn't you be able to run the tool 3 times in order to add three external ontologies? Maybe it is not so important to let the user do all three in one step? On the other hand, maybe it would not be too difficult to extend it so that it would do 3 at once.

**Review by sheridancbio — CHANGES_REQUESTED** [review]
Looks good .. but a few typos and a few suggested rewordings / renamings. Also, this PR should include the mapping script itself.

File: docs/OncoTree-Mapping-Tool.md
Comment (ritikakundra) [reply docs/OncoTree-Mapping-Tool.md:3]: I like Cross ontology. Will change to that

Comment (ritikakundra) [reply docs/OncoTree-Mapping-Tool.md:3]: I will change it to Ontology to Ontology Mapping tool to keep it in sync with oncotree to oncotree mappint

File: docs/OncoTree-Mapping-Tool.md
Comment (ritikakundra) [reply docs/OncoTree-Mapping-Tool.md:194]: I will ask @rmadupuri to add it

File: docs/OncoTree-Mapping-Tool.md
Comment (ritikakundra) [reply docs/OncoTree-Mapping-Tool.md:205]: @sheridancbio for the time being we are only allowing one source code. The input file contains the source code. We can modify here to say just add one source code. @rmadupuri can this be updated in the read me?

File: docs/OncoTree-Mapping-Tool.md
Comment (ritikakundra) [reply docs/OncoTree-Mapping-Tool.md:212]: Agreed

File: docs/OncoTree-Mapping-Tool.md
Comment (rmadupuri) [reply docs/OncoTree-Mapping-Tool.md]: @sheridancbio @ritikakundra added the script and renamed it to `ontology_to_ontology_mapping_tool.py`

File: docs/OncoTree-Mapping-Tool.md
Comment (rmadupuri) [reply docs/OncoTree-Mapping-Tool.md:205]: @sheridancbio, since we are currently allowing only one to one ontology mapping, if the source file has 2 or 3 of these fields the user will be asked to enter one ontology to map by.

File: docs/OncoTree-Mapping-Tool.md
Comment (rmadupuri) [reply docs/OncoTree-Mapping-Tool.md]: Agreed and updated per your suggestion.

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [reply docs/OncoTree-Mapping-Tool.md:205]: Ok @rmadupuri ... so the script will exit / fail and tell the user that only one ontology column can be present in the source file?

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:126]: @sheridancbio, heres's what the script does when it encounters more than one ontology in the source file.

File: docs/OncoTree-Mapping-Tool.md
Comment (rmadupuri) [reply docs/OncoTree-Mapping-Tool.md:205]: Yes @sheridancbio, just added another comment pointing to the block in code that handles this case.

File: scripts/ontology_to_ontology_mapping_tool/ontology_mappings.txt
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_mappings.txt:1]: I've scanned through this mappings file and I don't think there are any values for the ICDO_CODE column. Is that intentional?

Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_mappings.txt:1]: @sheridancbio yes. @ritikakundra will update the mappings soon as we get them.

File: docs/OncoTree-Mapping-Tool.md
Comment (sheridancbio) [inline docs/OncoTree-Mapping-Tool.md:194]: Users do not need to download the mapping file in order to run the tool; I would make that clear here with a short statement (The mapping file does not need to be downloaded to run the tool). I now realize that in order for the user to be able to use the tool, they will need to have an internet connection so that the script can load the mapping file from github. Maybe internet connections are now an assumed part of life ... so maybe we don't need to say anything about it. Hopefully the script will output a sensible message if a user tries to run it with no network access.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:1]: 2020

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:21]: are pandas and requests automatically built in to python3? or do they need to be installed? (I think maybe requests is not there by default). If either one is a prerequisite to running the script, we should mention it in the text on the Mapping Tool page and give an example of how to install it (pip3 install pandas .. maybe?)

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:29]: throughout this file .. add a space character after comma in argument lists / function calls

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:40]: why did we change "pandas" to "pd"? I think it would be more clear if we leave it as "pandas" unless everyone who uses pandas always sets it to "pd".

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:41]: This is not valid. This file contains "triplets" of the form "subject verb object". This approach takes all objects as valid oncotree codes, but sometimes objects are not .. such as with the triplet
`ONC000688	hasPrecursor	ONC000360`
This says that concept ONC000688 has a precursor concept ONC000360 ... and ONC000360 is not a valid oncotree code. Maybe you could only use triplets of the form (subject hasCode object) and then use the object.
Also .. there are some rescinded oncotree codes, such as RESCINDED_URI_PREVIOUSLY_USED_FOR_PMBL. These should never occur in user files.
Also .. there are some confusing cases in the history, where oncotree codes were changed and/or rescinded. So
ONC000376	hasCode	PTCL
ONC000379	hadCode	PTCLNOS
ONC000379	hasRevocation	ONC000376
so this will take both PTCL and PTCLNOS as valid oncotree codes ... but above you give a mapping for PTCL (which is no longer a current code) and do not give a mapping for PTCLNOS (which replaced PTCL) ... there are other cases where codes were rescinded like this. I'm not sure what scenarios we want to support, but it can get complicated.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:43]: add space after comma in tuple lists too

**Review by sheridancbio — CHANGES_REQUESTED** [review]
partial review (up to start of script)

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:130]: I see.. we are allowing the user to enter a choice manually to solve the problem. If this script is run as part of an automated pipeline by someone else ... then the automated pipeline will 'hang' / stall at this input statement I think. Maybe we should add a command line argument so that if a user is running in batch mode it will not try to ask the user for this input but will instead just fail at this point. Either that or we could add an "interactive" flag (-i // --interactive) which would enable the user to answer this question if it comes up. This is optional .. but if we do not do something like this, then we should mention in the README or on the Mapping Tools page that the script might ask for manual input when there are multiple source ontologies present, and so it is only intended to be run manually (and not as part of an automated pipeline).

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:136]: good check --- maybe another good check to add would be to see if the target ontology column already exists in the source file. The user might have a source of oncotree and a target of ICD-O ... but what would happen if the ICD-O column is already present in their source file? I suppose it would be overwritten? Maybe that is ok.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:156]: data is too general ... maybe oncotree_code_source_data ?

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:160]: there is one (old) oncotree code which had a lowercase letter ... we probably can ignore it because it is no longer current and we do not plan to allow lowercase letters again.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:163]: pandas has this merge utility ... nice

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:168]: we overwrite the old summary.html file I suppose .. which is fine.

**Review by sheridancbio — CHANGES_REQUESTED** [review]
part two of review.

Overall .. looks good 👍

File: docs/OncoTree-Mapping-Tool.md
Comment (rmadupuri) [reply docs/OncoTree-Mapping-Tool.md]: Updated the README to include the statement @sheridancbio

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: Updated to 2020

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:21]: Good point @sheridancbio.  I have added a Prerequisites section to readme with the requirements and how to install the libraries.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: Fixed

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: I have imported the pandas as "pd" only because its easier to write eachtime I need to use it :)

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: Thanks for pointing it out @sheridancbio. Updated the script to use only the records with `hasCode`.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: Done

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: That's right @sheridancbio. I have updated the script to read the source code as a command line argument instead of looking for it in the file. Updated the README as well.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:119]: If the source file already has the columns, ICD-O in your example, the common columns in the source file will get the suffix `_source_file` ICDO_source_file here.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: Updated

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:144]: Yes, there was one code SpCC. I guess its no more part of the tree? And all the ontologies were uppercase in the mapping file.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (rmadupuri) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:152]: I guess it's fine?

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:119]: ok, that seems reasonable. I guess if the user starts with ICDO column and then runs the script twice they may wind up with ICDO_source_file_source_file ... but we can leave it to them to clean up those columns I guess.

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:144]: 👍

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: ok

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: ok, looks good. There still may be some confusion and lack of support if they have not updated to the latest oncotree version .. but that is reasonable

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [reply scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py]: 👍  this works for everyone (even if it means a little extra typing)

File: scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py
Comment (sheridancbio) [inline scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py:127]: sorry .. one more change request ... this massage should be "does not contain the requested source ontology code column"

**Review by sheridancbio — CHANGES_REQUESTED** [review]

**Review by sheridancbio — APPROVED** [review]
nice work

**Pull request merged**
